packages/test-base/src/androidAndroidTest/kotlin/io/realm/kotlin/test/shared/CopyFromRealmTests.kt

```diff
@@ -0,0 +1,747 @@
+            nullableObject = topLevelObject
+            objectListField = realmListOf(topLevelObject)
+            objectSetField = realmSetOf(topLevelObject)
+        }
+
+        val unmanagedCopy = realm.writeBlocking {
+            copyToRealm(sample).copyFromRealm()
+        }
+
+        assertSame(unmanagedCopy, unmanagedCopy.nullableObject)
+        assertSame(unmanagedCopy, unmanagedCopy.objectListField.first())
+        assertSame(unmanagedCopy, unmanagedCopy.objectSetField.first())
+    }
+
+    @Test
+    fun objectGraphShareMemRefForSharedObjects() {
+        val sample = Sample().apply {
+            val shared = Sample().apply { stringField = "shared" }
+            nullableObject = shared
+            objectListField = realmListOf(shared)
+            objectSetField = realmSetOf(shared)
+        }
+        val unmanagedCopy = realm.writeBlocking {
+            val result = copyToRealm(sample)
+            val obj1 = result.nullableObject
+            val obj2 = result.objectListField.first()
+            val obj3 = result.objectSetField.first()
+            assertNotEquals(obj1, obj2) // Managed objects use default `equals`, i.e. compare mem ref.
+            assertNotEquals(obj2, obj3) // Managed objects use default `equals`, i.e. compare mem ref.
+            result.copyFromRealm()
+        }
+        realm.close()
+
+        // When copying RealmObjects back into memory, the same underlying data should result in the
+        // the same object in Kotlin, ie. they should share mem ref.
+        val objectCopy: Sample = unmanagedCopy.nullableObject!!
+        assertNotSame(objectCopy, sample.nullableObject)
+        assertEquals("shared", objectCopy.stringField)
+        assertSame(objectCopy, unmanagedCopy.objectListField.first())
+        assertSame(objectCopy, unmanagedCopy.objectSetField.first())
+    }
+
+    @Test
+    fun objectsAtDifferentVersionsAreDifferentAfterCopy() {
+        val sample = Sample().apply {
+            stringField = "v1"
+            intField = 3
+        }
+
+        val managedSample1 = realm.writeBlocking {
+            copyToRealm(sample)
+        }
+        val managedSample2 = realm.writeBlocking {
+            findLatest(managedSample1)!!.apply {
+                intField = 42
+            }
+        }
+        val o1p = (managedSample1 as RealmObjectInternal).io_realm_kotlin_objectReference!!.objectPointer
+        val o2p = (managedSample2 as RealmObjectInternal).io_realm_kotlin_objectReference!!.objectPointer
+        assertFalse(RealmInterop.realm_equals(o1p, o2p))
+
+        val unmanagedObjects = realm.copyFromRealm(listOf(managedSample1, managedSample2))
+        assertEquals(2, unmanagedObjects.size)
+        val unmanagedSample1 = unmanagedObjects.first()
+        val unmanagedSample2 = unmanagedObjects.last()
+
+        assertNotSame(unmanagedSample1, unmanagedSample2)
+        assertNotEquals(unmanagedSample1, unmanagedSample2)
+        assertEquals(3, unmanagedSample1.intField)
```

Comment:
Maybe also assert that `stringField == "v1"` to highlight that the two copies originate from the same object.